### ISSUE TITLE
Guidance on extending the action space with a new (mobile-base) channel for post-training on RoboCasa365

### ISSUE THREAD

**Issue by BigBen111** [sharinka0715/X-WAM#2]
Hi, thanks for the great work and for open-sourcing X-WAM!

We're post-training (SFT) X-WAM on **RoboCasa365** (Human300 split), which contains
**mobile manipulation** — many tasks require driving the robot base. Since X-WAM's
action space is end-effector-only (`action_dim=14`), we extended it with a base
channel. We'd appreciate your advice on the approach and on a training observation.

## What we did

**Action / proprio extension (cold start — the pretraining corpus is all fixed-base):**
- `action_dim`: 14 → **19** = `[left_eef(7), right_eef(7), base_motion(4), control_mode(1)]`
- `proprio_dim`: 16 → **23** = `[left(8), right(8), base_pos(3), base_quat(4)]`
- `control_mode ∈ {-1=arm, +1=base}`, trained as a continuous flow-matching dim,
  thresholded at 0 during eval.

**Grafting (`scripts/train_sft.py`):** we resize the 6 projection layers
(`action_encoder.0`, `action_decoder.2` w+b, `proprio_encoder.0`,
`proprio_decoder.2` w+b). For each, the rows/cols that **overlap** with the original
dims (the first 14 action dims / 16 proprio dims = the eef channels) are **copied
from your pretrained checkpoint**, and the **new base rows/cols are zero-init**
(function-preserving). The DiT backbone is unchanged (dim-parameterized). So the eef
side is warm-started and only the base side is cold-started.

**Loss:** masked MSE on the flow-matching velocity, with a **per-dimension** weight —
the 14 eef dims use `w=1`, the 5 new base dims use `w=2`:

$$
\mathcal{L}_{\text{action}} = \frac{\sum_{b,t,d}\ m_{b,t,d}\ w_d\ (\hat{v}_{b,t,d} - v_{b,t,d})^2}{\sum_{b,t,d}\ m_{b,t,d}}
$$

Here the square is element-wise per action dimension $d$ (not a vector norm over $d$) — 
that's what allows the per-dim weight $w_d$.  $m\in\{0,1\}$ the mask. The denominator is **unweighted**,
so this is a genuine 2× gradient on the base dims.

## Config

| | value |
|---|---|
| init | `sharinka0715/X-WAM-checkpoints` pretrained/ |
| training | full fine-tune (6.67B; T5+VAE frozen), DeepSpeed ZeRO-2 |
| lr / schedule | 1e-5, cosine→0 at 100k, warmup 1000 |
| batch | 4/GPU × 8 GPUs (A800), grad-checkpointing |
| ANS | `use_joint_distribution=true`, `clean_action_ratio=0.5`, `time_shifting=5.0` |
| data | Human300 (atomic + composite), ~32k episodes |

## What we observe

Post-training from your pretrained checkpoint, **`arm_action_loss` & `video` barely decreases 
from 0 to 20k steps.** All the other losses ( depth / proprio / base) look normal.

*(loss curves attached as a screenshot below)*

## Questions

1. Is our **graft + zero-init** approach for adding a new action channel the intended
   way to extend the action/proprio space, or would you recommend something else?
2. Are the **lr / schedule / steps** we're using appropriate for this post-training, or would
   you suggest different settings to get the arm loss to keep decreasing?
3. Did the **pretraining corpus** include any mobile-base / navigation data, and is
   SFT alone expected to be enough to learn a new base channel?

Thanks a lot!

In the screenshot, we further break down the action loss into arm_action_loss and base_action_loss, where base_action_loss measures the error of the mobile base actions.

<img width="1040" height="462" alt="Image" src="https://github.com/user-attachments/assets/c09e262a-c77f-41d3-a4bd-06617ceb2d5c" />
<img width="1047" height="477" alt="Image" src="https://github.com/user-attachments/assets/90c22e05-073f-44bc-a313-e19cd658521a" />
<img width="1572" height="471" alt="Image" src="https://github.com/user-attachments/assets/ac3e2b7f-7c26-45c2-a3f2-544ccd50ea73" />

**Comment by sharinka0715** [comment]
Thanks for your interest in our work!

Unfortunately, our work focuses only on fixed-base bimanual manipulation. We did not include any mobile-base data in the training set, nor have we systematically studied related settings. Therefore, it is difficult for me to give a definitive answer without running experiments. That said, from my perspective, even with your current extension approach, SFT should still be able to reduce these losses to some extent.

Regarding your questions:

1. I think your current grafting strategy is a reasonable and well-designed way to extend the action/proprioception space. It preserves the pretrained knowledge on the original action dimensions while introducing the new action channels. In fact, as shown in your results, the loss on the newly added base action dimensions decreases quite clearly.

2. I suspect the current batch size may be somewhat small, which could make optimization less stable or cause the model to get stuck. This might be one possible reason why the original arm-action loss is hard to reduce. With gradient checkpointing enabled, you may be able to further increase the batch size, for example to 8 or 16 per GPU, or use more GPUs if available. Increasing the learning rate moderately together with the larger effective batch size might also help.

3. The pretraining data does not contain any mobile-base or navigation data. My guess is that SFT alone should still be able to achieve a reasonably good result, but based on your current curves, there should still be room for further optimization.

Thanks again for your interest in our work, and I hope you can resolve this issue soon!

**Comment by BigBen111** [comment]
Thanks for the detailed reply!

**Comment by safe-sentinel-co** [comment]
did you figure anythign out? we are running into a similar issue where the action loss is flat and video loss doesn't budge much

**Comment by safe-sentinel-co** [comment]
@BigBen111 did you manage to fix the issue?
